### PR TITLE
test: Step-based checkpointing fixes

### PR DESCRIPTION
**Pull request by bogdansalyp**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/configs/llama3_2/1B_full_single_device.yaml [hunk 1]
```diff
@@ -35,19 +35,18 @@ tokenizer:
...
-  checkpoint_files: [
-    model.safetensors
-  ]
-  recipe_checkpoint: null
+  checkpoint_files: [model.safetensors]
   output_dir: ${output_dir}
   model_type: LLAMA3_2
+  keep_last_n_checkpoints: 2
 resume_from_checkpoint: False
+save_every_n_steps: 25
```
Comment (joecummings) [inline recipes/configs/llama3_2/1B_full_single_device.yaml:49]: These were just test values, can probably remove

File: recipes/configs/llama3_2/1B_full_single_device.yaml [hunk 2]
```diff
@@ -57,7 +56,7 @@ optimizer:
   lr: 2e-5
 loss:
   _component_: torchtune.modules.loss.LinearCrossEntropyLoss
-max_steps_per_epoch: null
+max_steps_per_epoch: 100
```
Comment (joecummings) [inline recipes/configs/llama3_2/1B_full_single_device.yaml:59]: Same here, just a test value

File: torchtune/training/checkpointing/_checkpoint_client.py [hunk 3]
```diff
@@ -182,20 +189,25 @@ def _save_checkpoint_async(
...
                 epoch=epoch,
                 save_async=True,
                 adapter_only=True,
+                step=training_progress.steps_run,
+                dir_prefix=dir_prefix,
             )
 
             if adapter_only:
+                save_path = dcp_saver.output_dir
+                if dir_prefix == "step":
```
Comment (joecummings) [inline torchtune/training/checkpointing/_checkpoint_client.py:203]: Might be worth a comment that this is fairly hacky b/c we need to infer BC with epochs and potentially could be refactored

File: torchtune/training/checkpointing/_checkpoint_client.py [hunk 4]
```diff
@@ -365,6 +408,9 @@ def save_checkpoint(
         adapter_config: Optional[dict[str, Any]] = None,
         adapter_only: bool = False,
         single_device: bool = False,
+        *,
+        full_tensors: bool = True,
```
Comment (joecummings) [inline torchtune/training/checkpointing/_checkpoint_client.py:412]: Maybe add a comment in the docstring explaining what full_tensors are and when you might want to use them.

File: torchtune/training/checkpointing/_checkpointer.py [hunk 5]
```diff
@@ -341,7 +344,7 @@ def save_checkpoint(
             _ = state_dict.pop(training.ADAPTER_KEY, None)
             _ = state_dict.pop(training.ADAPTER_CONFIG, None)
             output_path = Path.joinpath(
-                self._output_dir, RECIPE_STATE_DIRNAME, "recipe_state.pt"
+                self._output_dir, f"epoch_{epoch}", "recipe_state.pt"
```
Comment (joecummings) [inline torchtune/training/checkpointing/_checkpointer.py:347]: Why just epoch here?

**Review by joecummings — APPROVED** [review]
Mostly nits, just one question about "epoch" usage in checkpointer

File: torchtune/training/checkpointing/_checkpointer.py (hunk 5 above)
Comment (bogdansalyp) [reply torchtune/training/checkpointing/_checkpointer.py:347]: It's a debugger torchtune checkpointer, not a HF one
